### PR TITLE
[Feature] Add Flashinfer MoE Support for Compressed Tensor NVFP4

### PR DESCRIPTION
**Pull request by yewentao256**
## Purpose

Add Flashinfer Support for Compressed Tensor NVFP4

Code base refactor in https://github.com/vllm-project/vllm/pull/21631, just for easier review.

### Accuracy

```bash

lm_eval  --model vllm  --model_args "pretrained=nm-testing/Qwen3-30B-A3B-NVFP4,max_model_len=32768,enable_expert_parallel=True,enforce_eager=True"  --trust_remote_code  --tasks gsm8k  --num_fewshot 5  --batch_size auto
# flashinfer path
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8779|±  |0.0090|
|     |       |strict-match    |     5|exact_match|↑  |0.8704|±  |0.0093|
# cutlass path
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8779|±  |0.0090|
|     |       |strict-match    |     5|exact_match|↑  |0.8757|±  |0.0091|
```

### Performance

```bash
vllm bench throughput --model nm-testing/Qwen3-30B-A3B-NVFP4 --load-format dummy --input-len 100 --output-len 100 --trust_remote_code --enable-expert-parallel
# flashinfer
Throughput: 180.33 requests/s, 35996.82 total tokens/s, 18033.13 output tokens/s
Total num prompt tokens:  99615
Total num output tokens:  100000
# cutlass
Throughput: 160.02 requests/s, 31943.00 total tokens/s, 16002.30 output tokens/s
Total num prompt tokens:  99615
Total num output tokens:  100000
```

### CODE REVIEW

File: vllm/model_executor/layers/quantization/modelopt.py [hunk 1]
```diff
@@ -877,26 +878,18 @@ def __init__(self, quant_config: ModelOptNvFp4Config):
...
-                                 " quantization. Please use Blackwell and"
-                                 " above.")
+        if self.allow_flashinfer_cutlass:
+            logger.info_once(
+                "Using FlashInfer kernels for ModelOptNvFp4FusedMoE.")
+        else:
+            if envs.VLLM_USE_FLASHINFER_MOE_FP4:
+                logger.warning_once(
+                    "FlashInfer kernels unavailable for "
+                    "ModelOptNvFp4FusedMoE on current platform.")
```
[the remark this answers is not in the document]
Comment (yewentao256) [reply vllm/model_executor/layers/quantization/modelopt.py]: FIxed

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py [hunk 2]
```diff
@@ -271,6 +307,40 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
...
+                tp_size=moe.moe_parallel_config.tp_size,
+            )
+            return experts
+
+        # Native CUTLASS experts currently don't support DP
+        assert moe.dp_size > 1
+        logger.debug_once("Using CutlassExpertsFp4 (unsupported for DP)")
+        raise ValueError(
+            "CutlassExpertsFp4 doesn't support DP. Use flashinfer CUTLASS "
+            "Fused MoE backend instead (set VLLM_USE_FLASHINFER_MOE_FP4=1)")
```
[the remark this answers is not in the document]
Comment (yewentao256) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py]: Fixed

**Comment by yewentao256** [comment]
CC @mgoin

File: vllm/model_executor/layers/quantization/modelopt.py [hunk 3]
```diff
@@ -923,8 +905,6 @@ def select_gemm_impl(self, prepare_finalize,
                 tp_size=moe.moe_parallel_config.tp_size,
             )
         else:
-            assert moe.dp_size > 1
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/modelopt.py:952]: It shouldn't hurt to keep this assert

File: vllm/model_executor/layers/quantization/utils/flashinfer_fp4_utils.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    )
+
+
+def flashinfer_fp4_forward(
+    fused_experts: mk.FusedMoEModularKernel,
+    layer: torch.nn.Module,
+    x: torch.Tensor,
+    topk_weights: torch.Tensor,
+    topk_ids: torch.Tensor,
+    *,
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/utils/flashinfer_fp4_utils.py]: No need for this asterisk. Personally I think we should use named args for all

File: vllm/model_executor/layers/quantization/modelopt.py [hunk 5]
```diff
@@ -1191,49 +1161,15 @@ def apply(
...
-                    "Flashinfer CUTLASS Fused MoE not applicable!")
-
-            a1_gscale = layer.w13_input_scale_quant
-            a2_gscale = layer.w2_input_scale_quant
-            extra_expert_args = {
-                'g1_alphas': layer.g1_alphas,
-                'g2_alphas': layer.g2_alphas,
-                'out_dtype': x.dtype,
-                # Avoid confusion with a1_scale and a2_scale
-                # where are batch size related.
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/modelopt.py:1234]: Please keep the comments

File: vllm/model_executor/layers/quantization/modelopt.py [hunk 6]
```diff
@@ -1191,49 +1161,15 @@ def apply(
...
-                'local_tokens': x.shape[0],
-            }
-
-            out = self.fused_experts(
-                hidden_states=x,
-                w1=layer.w13_weight,
-                w2=layer.w2_weight,
-                topk_weights=topk_weights,
-                topk_ids=topk_ids,
-                inplace=False,  # TODO(shuw): fix later, now output is high prec
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/modelopt.py:1254]: Ditto, keep the comment

File: vllm/model_executor/layers/quantization/utils/flashinfer_fp4_utils.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        tp_rank=moe_parallel_config.tp_rank,
+        tp_size=moe_parallel_config.tp_size,
+    )
+    return mk.FusedMoEModularKernel(
+        FlashInferCutlassMoEPrepareAndFinalize(quant_dtype=torch.uint8),
+        experts,
+    )
+
+
+def flashinfer_fp4_forward(
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/utils/flashinfer_fp4_utils.py]: This name is vague, maybe `flashinfer_fp4_cutlass_moe_forward`

File: vllm/model_executor/layers/quantization/utils/flashinfer_fp4_utils.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+    w1, w3 = weight.split(half, dim=dim)
+    s1, s3 = scale.split(half, dim=dim)
+
+    return (torch.cat([w3, w1],
+                      dim=dim).contiguous(), torch.cat([s3, s1],
+                                                       dim=dim).contiguous())
+
+
+def build_flashinfer_kernel(
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/utils/flashinfer_fp4_utils.py]: `build_flashinfer_fp4_cutlass_moe[_kernel]`

File: vllm/model_executor/layers/quantization/utils/flashinfer_fp4_utils.py
Comment (mgoin) [inline vllm/model_executor/layers/quantization/utils/flashinfer_fp4_utils.py:1]: It is already under the utils folder, so we can call it `flashinfer_fp4_moe.py`

File: vllm/model_executor/layers/quantization/utils/flashinfer_fp4_utils.py [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+__all__ = [
+    "is_flashinfer_fp4_available",
+    "reorder_w1w3_to_w3w1",
+    "build_flashinfer_kernel",
+    "flashinfer_fp4_forward",
+]
+
+
+def is_flashinfer_fp4_available() -> bool:
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/utils/flashinfer_fp4_utils.py]: `is_flashinfer_fp4_cutlass_moe_available`

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py [hunk 10]
```diff
@@ -98,8 +104,30 @@ def get_moe_method(
...
+        if not self.cutlass_nvfp4_supported:
+            if is_fp4_marlin_supported():
+                self.use_marlin = True
+            else:
+                raise ValueError(
+                    "Current platform does not support NVFP4 quantization. "
+                    "Please use Blackwell and above.")
+
         self.group_size = 16
+        self.fused_experts = None  # type: ignore[assignment]
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py]: Do you think this could be in a shared util between modelopt and ct as well?

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py [hunk 11]
```diff
@@ -248,6 +284,39 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
         layer.w2_input_scale_quant = torch.nn.Parameter(
             (layer.w2_input_global_scale), requires_grad=False)
 
+    def maybe_swap_experts_impl(self, moe_parallel_config):
+        if not self.allow_flashinfer_cutlass:
+            return
+        self.fused_experts = build_flashinfer_kernel(moe_parallel_config)
+        logger.debug_once("FlashInferExperts (util)")
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py]: Put this log in the `build_flashinfer_kernel` call?

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py [hunk 12]
```diff
@@ -248,6 +284,39 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
...
+                tp_rank=moe.moe_parallel_config.tp_rank,
+                tp_size=moe.moe_parallel_config.tp_size,
+            )
+            return experts
+
+        # Native CUTLASS experts currently don't support DP
+        # TP case won't call this function
+        raise ValueError(
+            "CutlassExpertsFp4 doesn't support DP. Use flashinfer CUTLASS "
+            "Fused MoE backend instead (set VLLM_USE_FLASHINFER_MOE_FP4=1)")
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py]: Seems like this should be shared as well

File: vllm/model_executor/layers/quantization/modelopt.py (hunk 3 above)
Comment (yewentao256) [reply vllm/model_executor/layers/quantization/modelopt.py:952]: Fixed

File: vllm/model_executor/layers/quantization/modelopt.py (hunk 5 above)
Comment (yewentao256) [reply vllm/model_executor/layers/quantization/modelopt.py:1234]: Fixed

File: vllm/model_executor/layers/quantization/utils/flashinfer_fp4_utils.py
Comment (yewentao256) [reply vllm/model_executor/layers/quantization/utils/flashinfer_fp4_utils.py:1]: Fixed

File: vllm/model_executor/layers/quantization/utils/flashinfer_fp4_utils.py (hunk 9 above)
Comment (yewentao256) [reply vllm/model_executor/layers/quantization/utils/flashinfer_fp4_utils.py]: Fixed

File: vllm/model_executor/layers/quantization/utils/flashinfer_fp4_utils.py (hunk 8 above)
Comment (yewentao256) [reply vllm/model_executor/layers/quantization/utils/flashinfer_fp4_utils.py]: Fixed

File: vllm/model_executor/layers/quantization/utils/flashinfer_fp4_utils.py (hunk 7 above)
Comment (yewentao256) [reply vllm/model_executor/layers/quantization/utils/flashinfer_fp4_utils.py]: Fixed

File: vllm/model_executor/layers/quantization/utils/flashinfer_fp4_utils.py (hunk 4 above)
Comment (yewentao256) [reply vllm/model_executor/layers/quantization/utils/flashinfer_fp4_utils.py]: Fixed

File: vllm/model_executor/layers/quantization/modelopt.py (hunk 6 above)
Comment (yewentao256) [reply vllm/model_executor/layers/quantization/modelopt.py:1254]: Fixed

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py (hunk 10 above)
Comment (yewentao256) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py]: Sounds good, updated

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py (hunk 11 above)
Comment (yewentao256) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py]: Fixed

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py (hunk 12 above)
Comment (yewentao256) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py]: Updated

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py [hunk 13]
```diff
@@ -96,8 +101,14 @@ def get_moe_method(
 class CompressedTensorsW4A4MoeMethod(CompressedTensorsMoEMethod):
 
     def __init__(self):
-        self.use_marlin = not cutlass_fp4_supported()
+        from vllm.model_executor.layers.quantization.utils.nvfp4_support import (  # noqa: E501
+            detect_nvfp4_support)
+        _nvfp4 = detect_nvfp4_support(self.__class__.__name__, logger)
+        self.cutlass_nvfp4_supported = _nvfp4.cutlass_supported
+        self.allow_flashinfer_cutlass = _nvfp4.allow_flashinfer_cutlass
+        self.use_marlin = _nvfp4.use_marlin
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py]: Could we just pass in self to the function? Then the function could take care of assigning these variables as well as getting the class name. 
Also, I don't see why we should pass in the logger here and the function name should include moe if this is moe specific

Comment (yewentao256) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py]: Yeah, the origin version is detect_nvfp4_support(obj, logger), and we set everything inside the function, like `obj. allow_flashinfer_cutlass = True`. But after consideration, this won't be a good practice since class attributes should be set inside class so that it is easy to refer and read.

Comment (yewentao256) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py]: Logger and function name should include moe fixed, thanks!

**Review by mgoin — APPROVED** [review]
